Hmm - another oidc code flow failure, seeing
```
 Caused by: io.fabric8.maven.docker.access.hc.http.HttpRequestException: {"message":"received unexpected HTTP status: 502 Bad Gateway"} (Internal Server Error: 500)
```

there was another failure (with `Connection was closed`) earlier

CC @gsmet, Hi Guillaume, has something changed around the way Keycloak images are started ? Also CC Clement @cescoffier

@jasondlee thanks for the PR, LGTM, can you please update some test involving `@TestSecurity` to verify these attributes are available ? For example, one test I'm aware of is in `integration-tests/oidc-code-flow` - perhaps you can update the test endpoint there to echo back the configured attributes
thanks

--- a/integration-tests/elytron-resteasy/src/main/java/io/quarkus/it/resteasy/elytron/RootResource.java
+++ b/integration-tests/elytron-resteasy/src/main/java/io/quarkus/it/resteasy/elytron/RootResource.java
@@ -1,6 +1,10 @@
 package io.quarkus.it.resteasy.elytron;
 
+import java.util.Map;
+import java.util.stream.Collectors;
+
 import javax.annotation.security.RolesAllowed;
+import javax.inject.Inject;
 import javax.ws.rs.Consumes;
 import javax.ws.rs.GET;
 import javax.ws.rs.POST;
@@ -11,9 +15,12 @@ import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.SecurityContext;
 
 import io.quarkus.security.Authenticated;
+import io.quarkus.security.identity.SecurityIdentity;
 
 @Path("/")
 public class RootResource {
+    @Inject
+    SecurityIdentity identity;
 
     @POST
     @Consumes(MediaType.TEXT_PLAIN)
@@ -50,4 +57,17 @@ public class RootResource {
         return sec.getUserPrincipal().getName();
     }
 
+    @GET
+    @Path("/attributes")
+    @Authenticated
+    public String getAttributes() {
+        final Map<String, Object> attributes = identity.getAttributes();
+        if (attributes == null || attributes.isEmpty()) {
+            throw new RuntimeException("No attributes were specified");
+        }
+
+        return attributes.entrySet().stream()
+                .map(e -> e.getKey() + "=" + e.getValue())
+                .collect(Collectors.joining(","));
+    }
 }

--- a/integration-tests/elytron-resteasy/src/test/java/io/quarkus/it/resteasy/elytron/TestSecurityTestCase.java
+++ b/integration-tests/elytron-resteasy/src/test/java/io/quarkus/it/resteasy/elytron/TestSecurityTestCase.java
@@ -13,6 +13,7 @@ import org.junit.jupiter.params.provider.MethodSource;
 import org.junit.jupiter.params.provider.ValueSource;
 
 import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.security.SecurityAttribute;
 import io.quarkus.test.security.TestSecurity;
 
 @QuarkusTest
@@ -63,7 +64,7 @@ class TestSecurityTestCase {
 
     @Test
     @TestSecurity(user = "testUser", roles = "wrong")
-    void testGetWithTestUserwrongRole() {
+    void testGetWithTestUserWrongRole() {
         given()
                 .when()
                 .get("/user")
@@ -80,6 +81,17 @@ class TestSecurityTestCase {
                 .then()
                 .statusCode(200)
                 .body(is("testUser"));
+    }
+
+    @Test
+    @TestSecurity(user = "testUser", roles = "user", attributes = { @SecurityAttribute(key = "foo", value = "bar") })
+    void testAttributes() {
+        given()
+                .when()
+                .get("/attributes")
+                .then()
+                .statusCode(200)
+                .body(is("foo=bar"));
     }
 
     static Stream<Arguments> arrayParams() {

--- a/test-framework/security/src/main/java/io/quarkus/test/security/QuarkusSecurityTestExtension.java
+++ b/test-framework/security/src/main/java/io/quarkus/test/security/QuarkusSecurityTestExtension.java
@@ -3,6 +3,7 @@ package io.quarkus.test.security;
 import java.lang.reflect.Method;
 import java.util.Arrays;
 import java.util.HashSet;
+import java.util.stream.Collectors;
 
 import javax.enterprise.inject.spi.CDI;
 
@@ -54,10 +55,16 @@ public class QuarkusSecurityTestExtension implements QuarkusTestBeforeEachCallba
                     throw new RuntimeException("Cannot specify roles without a username in @TestSecurity");
                 }
             } else {
-                QuarkusSecurityIdentity user = QuarkusSecurityIdentity.builder()
+                QuarkusSecurityIdentity.Builder user = QuarkusSecurityIdentity.builder()
                         .setPrincipal(new QuarkusPrincipal(testSecurity.user()))
-                        .addRoles(new HashSet<>(Arrays.asList(testSecurity.roles()))).build();
-                CDI.current().select(TestIdentityAssociation.class).get().setTestIdentity(user);
+                        .addRoles(new HashSet<>(Arrays.asList(testSecurity.roles())));
+
+                if (testSecurity.attributes() != null) {
+                    user.addAttributes(Arrays.stream(testSecurity.attributes())
+                            .collect(Collectors.toMap(s -> s.key(), s -> s.value())));
+                }
+
+                CDI.current().select(TestIdentityAssociation.class).get().setTestIdentity(user.build());
             }
         } catch (Exception e) {
             throw new RuntimeException("Unable to setup @TestSecurity", e);

--- a/test-framework/security/src/main/java/io/quarkus/test/security/SecurityAttribute.java
+++ b/test-framework/security/src/main/java/io/quarkus/test/security/SecurityAttribute.java
@@ -1,0 +1,13 @@
+package io.quarkus.test.security;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({})
+public @interface SecurityAttribute {
+    String key();
+
+    String value();
+}

--- a/test-framework/security/src/main/java/io/quarkus/test/security/TestSecurity.java
+++ b/test-framework/security/src/main/java/io/quarkus/test/security/TestSecurity.java
@@ -24,4 +24,5 @@ public @interface TestSecurity {
      */
     String[] roles() default {};
 
+    SecurityAttribute[] attributes() default {};
 }

